Add documentation for Preview Features (#1464)
dapr/docs: daprdocs/content/en/operations/configuration/preview-features.md

---
type: docs
title: "How-To: Enable preview features"

> linkTitle: "Preview Features"

Reviewer: ```suggestion
linkTitle: "Preview features"
```
https://docs.dapr.io/contributing/contributing-docs/#style-and-tone

weight: 7000
description: "How to specify and enable preview features"
---

## Overview

> Some features in dapr are considered experimental when they are first released. These features require explicit opt-in in order to be used. The opt-in is specified in dapr's configuration.

Reviewer: ```suggestion
Some features in Dapr are considered experimental when they are first released. These features require explicit opt-in in order to be used. The opt-in is specified in Dapr's configuration.
```
https://docs.dapr.io/contributing/contributing-docs/#style-and-tone

> Currently, preview features are enabled on a per application basis. A global scope may be introduced in the future should there be a use case for it.

Reviewer: ```suggestion
Currently, preview features are enabled on a per application basis when running on Kubernetes. A global scope may be introduced in the future should there be a use case for it.
```

> ### Current Preview Features

Reviewer: ```suggestion
### Current preview features
```
https://docs.dapr.io/contributing/contributing-docs/#style-and-tone

Below is a list of existing preview features:

> - Actor Reentrancy

Reviewer: Do we have a doc on actor reentrancy (what it is, how to use it)?

Reviewer: If so we should link to it here, if not we should add a page

Author: Looks like an issue for it was missed. I'll tackle that today and include the reference between the two in whichever PR is approved/merged later.

https://github.com/dapr/docs/issues/1482

Reviewer: We'll get this in within the other PR so I can merge this and the other build will complete

> ## Configuration Elements

Reviewer: ```suggestion
## Configuration properties
```

The `features` section under the `Configuration` spec contains the following properties:

| Property       | Type   | Description |
|----------------|--------|-------------|
|name|string|The name of the preview feature that will be enabled/disabled
|enabled|bool|Boolean specifying if the feature is enabled or disabled

> ## Enabling a Preview Feature

Reviewer: ```suggestion
## Enabling a preview feature
```
https://docs.dapr.io/contributing/contributing-docs/#style-and-tone

> As stated above, preview features are specified in the configuration. Here is an example of a full configuration that contains multiple features:

Reviewer: ```suggestion
Preview features are specified in the configuration. Here is an example of a full configuration that contains multiple features:
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: featureconfig
spec:
  tracing:
    samplingRate: "1"
    zipkin:
      endpointAddress: "http://zipkin.default.svc.cluster.local:9411/api/v2/spans"
  features:
    - name: Feature1
      enabled: true
    - name: Feature2
      enabled: true
```

### Standalone

> To enabled a preview feature when running dapr locally, the configuration can be provided in two separate ways.
> 
> The primary way to provide the configuration is by specifying it while starting an application with the CLI.

Reviewer: ```suggestion
To enable preview features when running Dapr locally, either update the default configuration or specify a separate config file using `dapr run`.

The default Dapr config is created when you run `dapr init`, and is located at:
- Windows: `%USERPROFILE%\.dapr\config.yaml`
- Linux/macOS: `~/.dapr/config.yaml`

Alternately, you can update preview features on all apps run locally by specifying the `--config` flag in `dapr run` and pointing to a separate Dapr config file:
```

```bash
dapr run --app-id myApp --config ./previewConfig.yaml ./app
```

> Should every app require the preview feature the configuration that dapr generates during `dapr init` can be modified. Doing this removes the need to specify `--config` in the `run` command as the configuration generated by dapr is used by default. Note, modifying this configuration by hand can be dangerous as an error in it would affect all applications that are started after the modification.

Reviewer: ```suggestion
```

### Kubernetes

> In kubernetes, the configuration must be provided via a configuration component. Using the same configuration as above, apply it via `kubectl`.

Reviewer: ```suggestion
In Kubernetes mode, the configuration must be provided via a configuration component. Using the same configuration as above, apply it via `kubectl`:
```

```bash
kubectl apply -f previewConfig.yaml
```

> This can then be referenced in any application by modifying the application's configuration to reference that specific configuration component via the `dapr.io/config` element. For example:

Reviewer: ```suggestion
This configuration component can then be referenced in any application by modifying the application's configuration to reference that specific configuration component via the `dapr.io/config` element. For example:
```
